simple red light classifier with opencv

pkorivi opened the pull request:
Implemented simple red light classifier and interface to change the c…

Commit:
Implemented simple red light classifier and interface to change the classifier

Commit:
minor refactoring

ser94mor commented:
This pull request **introduces 3 alerts** when merging 0913b6edc55fab92c27814b346bfe6d86c99cabc into 8c3fdb3c728e35cff023e560a5c1039e26697909 - [view on LGTM.com](https://lgtm.com/projects/g/aoool/self-driving-car-using-ros/rev/pr-f8a3ef31d3b834d61fb1d304734dab2974c60a19)

**new alerts:**

* 2 for Unused local variable
* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

ser94mor commented on ros/src/tl_detector/light_classification/tl_classifier.py:
```diff
@@ -1,11 +1,53 @@
+                if len(approxcontour)>5:
+                    red_count += 1
+
+        # rospy.loginfo("red :  %d",red_count)
```
Please, remove the commented debug code.

pkorivi replied:
removed

pkorivi replied:
I also updated the false classification, it was due to detection of lights on the right side. I updated the thresholds. I didnt encounter the other situation where car doesnt stop on red lights in simulator. Try with this implementation once.

ser94mor commented on ros/src/tl_detector/light_classification/tl_classifier.py:
```diff
@@ -1,11 +1,53 @@
 from styx_msgs.msg import TrafficLight
+import cv2
+import numpy as np
+import rospy
```
`rospy` package is not used here. Please, remove.

pkorivi replied:
removed

ser94mor commented on ros/src/tl_detector/light_classification/tl_classifier.py:
```diff
@@ -1,11 +1,53 @@
+                approxcontour = cv2.approxPolyDP(contour, 0.01 * arclength, True)
+                #Find the coordinates of the polygon with respect to he camera frame in pixels
+                rect_cordi = cv2.minAreaRect(contour)
+                obj_x = int(rect_cordi[0][0])
```
The values assigned to local variables `obj_x` and `obj_y` are never used.

pkorivi replied:
The idea was to check where the lights are in the image but seemed kind of difficult to say, car pose may vary. In general, cases, it should be straight but as it was working without this implementation also I didn't use them. Removed this part of code

ser94mor commented on ros/src/tl_detector/tl_detector.py:
```diff
@@ -138,16 +138,18 @@ def get_light_state(self, light):
+            self.prev_light_loc = None
+            return False
+
+        cv_image = self.bridge.imgmsg_to_cv2(self.camera_image, "bgr8")
```
There is no such variable property. Replace with `self.camera_image_msg`.

pkorivi replied:
updated, copy paste issue :-( I was testing on my repository and looks like I pushed this by testing other repo.

ser94mor reviewed (changes requested):
On my machine, it falsely classifies green lights as red, which leads to 2x time decrease in speed, though it does not lead to a complete stop. The car also does not stop on red lights. It decreases the speed down to 1 m/h and keeps moving with this low speed beyond the stop line when there is a red traffic light.

Commit:
updated the thresholds and modified the message name

Commit:
updated review comments

ser94mor commented:
Works with almost all traffic lights except for the last one on the Highway track. 
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/2496083/51428650-1e519780-1c17-11e9-8f4e-235e1425aea5.gif)

Commit:
updated thresholds for stopping and red light size

ser94mor commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -113,7 +113,7 @@ def decelerate_waypoints(self, waypoints, closest_idx):
 
             # We subtract 2 below to stop front of the car in front of the stop line.
```
Please, update the comment to correspond to the current code state.

pkorivi committed:
Update waypoint_updater.py

Updated the comment in line with the implementation from 2 to 4 for number of stopping ahead wapoints

Elgeweily commented on ros/src/tl_detector/light_classification/tl_classifier.py:
```diff
@@ -1,11 +1,47 @@
+            if 18 < contourarea < 900: #Discard contours with a too large area as this may just be noise
+                arclength = cv2.arcLength(contour, True)
+                approxcontour = cv2.approxPolyDP(contour, 0.01 * arclength, True)
+                #Check for Square
```
Check for circle?

pkorivi replied:
It checks for both circle and length. 18<contourarea<900 checks for the number of pixels. This area of the polygon basically. and the len(approxcontours) checks for the type of contour. Though its not playing major role in our case as threshold is set low.

Elgeweily commented on ros/src/tl_detector/light_classification/tl_classifier.py:
```diff
@@ -1,11 +1,47 @@
+        mask2 = cv2.inRange(hsv, lower_red2, upper_red2)
+        red_img = cv2.addWeighted(mask1,1.0,mask2,1.0,0)
+
+        im, contours, hierarchy = cv2.findContours(red_img,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE)
```
try to check for a minimum number of red pixels, instead of the shape of a red region
